utils-intervalrdd/src/main/scala/org/bdgenomics/intervaltree/IntervalTree.scala

```diff
@@ -0,0 +1,402 @@
+/**
+ * Licensed to Big Data Genomics (BDG) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The BDG licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bdgenomics.intervaltree
+
+import scala.reflect.ClassTag
+import scala.collection.mutable.ListBuffer
+
+/*
+ * Interval tree that supports two dimensional range searches over keyed nodes
+ * in tree
```

Comment:
Could you describe this implementation more formally?  See e.g. https://en.wikipedia.org/wiki/Interval_tree